### PR TITLE
Rework status codes

### PR DESCRIPTION
**Pull request by mikekistler**
This PR illustrates the change I am proposing in #197.  Some work is still needed to polish this up but I think the key features are there -- enough to illustrate the proposal.

Key elements:
- the http library now validates and processes the target property of the `@statusCode` decorator.  The property value must be a numeric or string literal or union of numeric or string literals which are three digit codes starting with 1-5.
- the openapi library implements a new decorator `@default` (might want a different name) to specify that the decorated model should be the "default" response of an operation.

I did not implement decorators for "2XX", "4XX", or "5XX" response codes, but these would be very similar to the `@default` decorator.

### CODE REVIEW

**Comment by nguerrera** [comment]
> `@default` (might want a different name)

Maybe `@defaultResponse`?

**Comment by nguerrera** [comment]
This looks good to me. I presume there will be fallout in the closed repo if we merge this, right? Would you be able to prep the reaction ahead of time so we can review that? Last time we moved around @statusCode's in standard lib, the flow to closed repo was blocked for a while. You should be able to make parallel changes following https://github.com/Azure/cadl-azure/blob/main/CONTRIBUTING.md#making-a-cross-cutting-change-across-both-repos

File: packages/openapi3/test/test-return-types.ts [hunk 1]
```diff
@@ -353,13 +328,9 @@ describe("openapi3: return types", () => {
     `
     );
     strictEqual(diagnostics.length, 3);
```
Comment (timotheeguerin) [inline packages/openapi3/test/test-return-types.ts:330]: you can use the new `expectDiagnostics` util function to pass the list of diagnostic that it should expect

File: packages/openapi/src/decorators.ts [hunk 2]
```diff
@@ -29,3 +29,22 @@ export function $extension(program: Program, entity: Type, extensionName: string
 export function getExtensions(entity: Type): ReadonlyMap<string, any> {
   return openApiExtensions.get(entity) ?? new Map<string, any>();
 }
+
+// The @default decorator can be applied to a model. When that model is used
+// as the return type of an operation, this return type will be the default response.
+const defaultResponses = new Set<Type>();
```
Comment (timotheeguerin) [inline packages/openapi/src/decorators.ts:35]: Should use the program to keep state not in global objects. 
```ts
program.stateSet(uniqueSymbolForThisDecorator)
```

File: packages/openapi/src/decorators.ts [hunk 3]
```diff
@@ -29,3 +29,22 @@ export function $extension(program: Program, entity: Type, extensionName: string
...
+export function $default(program: Program, entity: Type) {
+  if (entity.kind !== "Model") {
+    reportDiagnostic(program, {
```
Comment (timotheeguerin) [inline packages/openapi/src/decorators.ts:38]: Depending on if this get merged first or the decorator test/validation https://github.com/microsoft/cadl/pull/236 there is a `validateDecoratorTarget` util function that does that and emit the diagnostic for you. 
"decorator-wrong-type" has also been removed from library

**Review by timotheeguerin — COMMENTED** [review]
Just a few nits

**Comment by mikekistler** [comment]
I've addressed @timotheeguerin 's comments, changed the decorator name to `@defaultResponse`, and rebased onto the current main.

I did not change the `@error` decorator name since I think we are still deciding if that should move to the Azure library.

I'll try to have the changed ready in cadl-autorest by tomorrow at the latest.

**Review by markcowl — APPROVED** [review]

File: packages/openapi3/src/openapi.ts [hunk 4]
```diff
@@ -450,95 +458,21 @@ function createOAPIEmitter(program: Program, options: OpenAPIEmitterOptions) {
...
-        return "Conflict";
-      case "412":
-        return "Precondition Failed";
-      case "5XX":
-        return "Server Error";
-      case "default":
-        return "An unexpected error response";
-    }
-    // We might want to throw here -- rather than giving some default
-    return "A successful response";
```
Comment (mikekistler) [inline packages/openapi3/src/openapi.ts:541]: I moved the status code description method to the http library and improved the handling of codes not explicitly listed.

**Comment by mikekistler** [comment]
@markcowl I addressed your comments.  Would you mind taking one more look for I merge?

**Review by markcowl — APPROVED** [review]

**Pull request merged**